drivers/nrf24l01p_ng/nrf24l01p_ng_netdev.c

```diff
@@ -0,0 +1,636 @@
+}
+
+static
+int _assert_awake(const nrf24l01p_ng_t *dev)
+{
+    uint8_t config = 0;
+    nrf24l01p_ng_read_reg(dev, NRF24L01P_NG_REG_CONFIG, &config, 1);
+    return config & NRF24L01P_NG_FLG_PWR_UP;
+}
+
+static
+netopt_state_t _state_to_netif(nrf24l01p_ng_state_t state)
+{
+    if (state == NRF24L01P_NG_STATE_POWER_DOWN) {
+        return NETOPT_STATE_SLEEP;
+    }
+    if (state == NRF24L01P_NG_STATE_STANDBY_1) {
+        return NETOPT_STATE_STANDBY;
+    }
+    if (state == NRF24L01P_NG_STATE_STANDBY_2) {
+        return NETOPT_STATE_TX;
+    }
+    if (state == NRF24L01P_NG_STATE_TX_MODE) {
+        return NETOPT_STATE_TX;
+    }
+    if (state == NRF24L01P_NG_STATE_RX_MODE) {
+        return NETOPT_STATE_RX;
+    }
+    return NETOPT_STATE_OFF; /* error */
+}
+
+nrf24l01p_ng_state_t _state_from_netif(netopt_state_t state)
+{
+    if (state == NETOPT_STATE_SLEEP) {
+        return NRF24L01P_NG_STATE_POWER_DOWN;
+    }
+    if (state == NETOPT_STATE_STANDBY) {
+        return NRF24L01P_NG_STATE_STANDBY_1;
+    }
+    if (state == NETOPT_STATE_TX) {
+        return NRF24L01P_NG_STATE_TX_MODE;
+    }
+    if (state == NETOPT_STATE_RX) {
+        return NRF24L01P_NG_STATE_RX_MODE;
+    }
+    return NRF24L01P_NG_STATE_UNDEFINED;
+}
+
+static
+void nrf24l01p_ng_irq_handler(void *_dev)
+{
+    nrf24l01p_ng_t *dev = (nrf24l01p_ng_t *)_dev;
+    /* Once the IRQ pin has triggered,
+       do not congest the thread´s
+       message queue with IRQ events */
+    gpio_irq_disable(dev->params.pin_irq);
+    DEBUG("[nrf24l01p_ng] IRQ\n");
+    netdev_trigger_event_isr((netdev_t *)dev);
+}
+
+static
+int _init(netdev_t *netdev)
+{
+    nrf24l01p_ng_t *dev = (nrf24l01p_ng_t *)netdev;
+    if (dev->params.config.cfg_data_rate >= NRF24L01P_NG_RF_DR_NUM_OF ||
+        dev->params.config.cfg_crc == NRF24L01P_NG_CRC_0BYTE ||
+        dev->params.config.cfg_channel >= NRF24L01P_NG_NUM_CHANNELS) {
+        return -ENOTSUP;
+    }
+    if (spi_init_cs(dev->params.spi, dev->params.pin_cs) != SPI_OK) {
+        DEBUG("[nrf24l01p_ng] _init(): spi_init_cs() failed\n");
+        return -EIO;
+    }
+    if (gpio_init(dev->params.pin_ce, GPIO_OUT) < 0) {
+        DEBUG("[nrf24l01p_ng] _init(): gpio_init() failed\n");
+        return -EIO;
+    }
+    gpio_clear(dev->params.pin_ce);
+    if (nrf24l01p_ng_acquire(dev) < 0) {
+        DEBUG("[nrf24l01p_ng] _init(): nrf24l01p_ng_acquire() failed\n");
+        return -EIO;
+    }
+    if (dev->state != NRF24L01P_NG_STATE_POWER_DOWN) {
+        nrf24l01p_ng_transition_to_power_down(dev);
+    }
+    nrf24l01p_ng_flush_tx(dev);
+    nrf24l01p_ng_flush_rx(dev);
+    uint8_t aw = NRF24L01P_NG_ADDR_WIDTH;
+    uint8_t bc[] = NRF24L01P_NG_BROADCAST_ADDR;
+    luid_base(dev->urxaddr.rxaddrpx.rx_p0, aw);
+    do {
+        /* "The LSByte must be unique for all six pipes" [datasheet p.38] */
+        luid_get(&dev->urxaddr.rxaddrpx.rx_p0[aw - 1], 1);
+    } while (dev->urxaddr.rxaddrpx.rx_p0[aw - 1] == bc[aw - 1]);
```

Comment:
> are there some clones that don't work quite right

Yes, there are. And even the  original seems to break easily. At least 2 of mine stopped to function properly out of nothing.
My modules look like:

![nrf24l01p](https://user-images.githubusercontent.com/15147337/81904102-16172b80-95c3-11ea-8568-eb4f5466042a.jpg)

The `TX_DS` interrupt occures if you send, and get an ACK or if you send and you did not request an ACK. If you don´t get an ACK, then a `MAX_RT` interrupt triggers.

From your previous output:
```
> 2020-05-04 19:34:02,988 #  [nrf24l01p_ng] Sending 4 bytes
2020-05-04 19:34:03,002 # [nrf24l01p_ng] IRQ
```
, I conclude that either `TX_DS` or `MAX_RT` has triggered.

When I look closer, I see:
```
2020-05-04 19:34:02,941 # DYNDP          [DPL_P5 0 DPL_P4 0 DPL_P3 0 DPL_P2 0 DPL_P1 0 DPL_P0 0]
2020-05-04 19:34:02,945 # FEATURES       [EN_DPL 0 EN_ACK_PAY 0 DYN_ACK 0]
```

This tells me that automatic ACK´s and dynamic payload lengths are disabled. Those are features of nrf24l01+ (Enhanced ShockBurst), which the nrf24l01 (ShockBurst) does not provide.

I think you bought the predecessor module (nrf24l01), which does only support static payload lengths of 32 byte and does not send ACK´s.

The nrf24l01+ is backwards compatible, if you disable automatic ACK, CRC, retransmissions, and dynamic payload and you cannot send faster than 250 kbps. I did not consider this in the new driver version.

Reply:
> 2020-05-14 00:11:19,739 # REG_DYNPD: 
2020-05-14 00:11:19,741 # 0x1c returned: 00000000
2020-05-14 00:11:19,742 # 
2020-05-14 00:11:19,742 # REG_FEATURE: 
2020-05-14 00:11:19,745 # 0x1d returned: 00000000

Here as well.

Maybe try to send exactly 32 bytes with `txtsnd`, then you could maybe see something on the other end. But I include the source address and the length now (6 bytes), so there are only 26 bytes left.

Reply:
> I think you bought the predecessor module (nrf24l01), which does only support static payload lengths of 32 byte and does not send ACK´s.

> Maybe try to send exactly 32 bytes with `txtsnd`, then you could maybe see something on the other end. But I include the source address and the length now (6 bytes), so there are only 26 bytes left.

Indeed!

```
2020-05-14 10:38:34,862 # txtsnd 4 A4:DF:DD:73:F5 012345678901234567892345
> 2020-05-14 10:38:34,863 #  [nrf24l01p_ng] Sending 31 bytes
```

and on the other end

```
2020-05-14 10:38:34,881 # [nrf24l01p_ng] Return exact frame length
2020-05-14 10:38:34,884 # [nrf24l01p_ng] Handle received frame
2020-05-14 10:38:34,885 # Rx frame
2020-05-14 10:38:34,901 # 0xA4 0xDF 0xDD 0x73 0xF5 0x63 0x05 0x7C 0x26 0x67 0x15 0x36 0x30 0x31 0x32 0x33 0x34 0x35 0x36 0x37 0x38 0x39 0x30 0x31 0x32 0x33 0x34 0x35 0x36 0x37 0x38 0x39 0x31 0x32 0x33 0x34 0x35 
2020-05-14 10:38:34,905 # [nrf24l01p_ng] Received frame length: 37
2020-05-14 10:38:34,910 # [nrf24l01p_ng] _adpt_recv: Invalid source address length
```

Reply:
Actually I can send smaller frames too:

```
2020-05-14 10:41:30,139 # txtsnd 4 A4:DF:DD:73:F5 1
2020-05-14 10:41:30,140 #  [nrf24l01p_ng] Sending 7 bytes
```

```
2020-05-14 10:41:30,157 # [nrf24l01p_ng] Return exact frame length
2020-05-14 10:41:30,160 # [nrf24l01p_ng] Handle received frame
2020-05-14 10:41:30,161 # Rx frame
2020-05-14 10:41:30,178 # 0xA4 0xDF 0xDD 0x73 0xF5 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0xBF 0x87 0x05 0x7C 0x26 0x67 0x15 0x36 0x31 
2020-05-14 10:41:30,181 # [nrf24l01p_ng] Received frame length: 37
2020-05-14 10:41:30,186 # [nrf24l01p_ng] _adpt_recv: Invalid source address length
```

Reply:
`2020-05-14 10:38:34,910 # [nrf24l01p_ng] _adpt_recv: Invalid source address length
`
Does this happen all the time? If not, I would blame it on possible bit errors, because nrf24l01 lacks CRC capability.

Reply:
Yes this is happening all the time - that's why the packet never gets handet up the the application / pktdump

Reply:
The frame seems to get padded with zeros if it is not 32 bytes. Actually I would expect it to work if the payload is 32 bytes. Then it should arrive as:
```
frame:
[D0, D1, D2, D3, D4, 0x05, S0, S1, S2, S3, S4, P0, P1, P2, P3, ..., P25]
D: Destination Byte
S: Source Byte
P: Payload Byte
And 0x05 is the source address length.
```
The bytes D0, ..., D4 are not actually sent, but appended by the `_recv` routine.
Howerver, I have not tested with your device, so I am not sure what happens.

Reply:
>  because nrf24l01 lacks CRC capability
Sorry, just want to correct this: It supports CRC but it is optional